### PR TITLE
DOC Improves shape documentation for *Flatten

### LINKED ISSUE #60841: [doc] nn.Unflatten output size is wrong
## 📚 Documentation

very clearly wrong. https://pytorch.org/docs/stable/generated/torch.nn.Unflatten.html

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @brianjo @mruberry @albanD

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/pytorch/pytorch/issues/60841

### CODE REVIEW

File: torch/nn/modules/flatten.py [hunk 1]
```diff
@@ -57,8 +57,8 @@ class Unflatten(Module):
       (tuple of `(name, size)` tuples) for `NamedTensor` input.
 
     Shape:
-        - Input: :math:`(N, *dims)`
-        - Output: :math:`(N, C_{\text{out}}, H_{\text{out}}, W_{\text{out}})`
+        - Input: :math:`(*)`. Input can be any shape.
+        - Output: :math:`(*, *unflattened\_size, *)`.
```
Comment (jbschlosser) [inline torch/nn/modules/flatten.py:61]: Any way to indicate that `*unflattened_size` is expanded at `dim`? Maybe something like `(*, dim, *)` before and `(*, *unflattened\_size, *)` after (where `*` represents any number of dimensions, including none)- wdyt?

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Mind making the analogous change in `Flatten` for consistency?

File: torch/nn/modules/flatten.py (hunk 1 above)
Comment (thomasjpfan) [reply torch/nn/modules/flatten.py:61]: `(*, dim, *)` would be inconsistent with how the size is normally specified in other Shape descriptions.

What do you think of using `dim` as a subscript: `(*, S_{dim}, *)`, where `S_{dim}` is the "size at dimension `dim`".

Comment (jbschlosser) [reply torch/nn/modules/flatten.py:61]: Cool, I agree with you that using `dim` as a subscript is more accurate. We could even explicitly point out that `S_{dim} = \prod *unflattened\_size` must hold - thoughts on that?

File: torch/nn/modules/flatten.py [hunk 2]
```diff
@@ -10,8 +10,8 @@ class Flatten(Module):
     Flattens a contiguous range of dims into a tensor. For use with :class:`~nn.Sequential`.
 
     Shape:
-        - Input: :math:`(N, *dims)`
-        - Output: :math:`(N, \prod *dims)` (for the default case).
+        - Input: :math:`(*, S_{start}, *, S_{end}, *)`
+        - Output: :math:`(*, \prod_{i=start}^{end} S_{i}, *)` (for the default case).
```
Comment (jbschlosser) [inline torch/nn/modules/flatten.py:14]: nit: I don't think "(for the default case)" is needed now since the more precise edit should cover all cases I believe

**Review by jbschlosser — COMMENTED** [review]
Looking good! Not to spend too much time on this, but since a bunch of doc updates are coming down the pipe soon, probably good to get a consistent set of rules in place. What are your opinions on:
* Defining explicitly what `S_{i}` represents (e.g. `where S_{i} is the size at dimension i`) - does this add anything or is it obvious?
* Using `N_{i}` instead of `S_{i}` here - does `N_{i}` fit with other docs or is it confusing?
* Defining explicitly what `*` represents - should we do this in the shape section of every doc?
* Explicitly indicating that the rest of output matches the input shape like the `Linear` docs do - does this add anything or is it obvious?
* Maybe there should be a section somewhere in the PyTorch docs defining the notation used in Shape sections? vs. defining everything in-place.

I guess in general there is a spectrum of notational rigor from loose + possibly implicit to tightly defined + explicit, with the end goal of having docs that are readable, understandable, and unambiguous to at least a reasonable extent. I think the strategy up until now has been taking it on a case-by-case basis, and this unfortunately has lead to some inconsistencies throughout the docs

**Comment by thomasjpfan** [comment]
> Looking good! Not to spend too much time on this, but since a bunch of doc updates are coming down the pipe soon, probably good to get a consistent set of rules in place. 

I think it's good to discuss now and spend a little time on this. Looks like we are going to end up working on https://github.com/pytorch/pytorch/issues/59566 for the "Shape" sections, while updating for no-batch.

> Defining explicitly what S_{i} represents (e.g. where S_{i} is the size at dimension i) - does this add anything or is it obvious?

I prefer explicit standing it.

> Using N_{i} instead of S_{i} here - does N_{i} fit with other docs or is it confusing?

`N` is commonly used for "batch". But as we move forward with "no batch", we would be consistently removing `N` and replacing it with a `*`. This means `N_{i}` can be used. I proposed "S" because of "size" and "tensor.size(1)".

> Defining explicitly what * represents - should we do this in the shape section of every doc?

I think we need to do this. Sometimes `*` can mean "any number of dimensions". But other times it could mean "only none or 1". For example `ReflectionPad2d` would be `(*, C, H_{in}, W_{in})` where `*` is one dimension or none.

> Maybe there should be a section somewhere in the PyTorch docs defining the notation used in Shape sections? vs. defining everything in-place.

Having a section defining the notation is good to have for developers regardless of if we define it in place. I do like having definitions in-place for users.

Overall I think the "Shape" section should be more on the explicit slide.

**Comment by thomasjpfan** [comment]
> Explicitly indicating that the rest of output matches the input shape like the Linear docs do - does this add anything or is it obvious?

For the Linear case, I think it is obvious because of how the `*` lines up. I think we can restrict `*` to only be used to be at the start and ends?

Something like this:

Screen Shot 2021-07-01 at 12 17 24 PM

**Comment by jbschlosser** [comment]
> For the Linear case, I think it is obvious because of how the `*` lines up. I think we can restrict `*` to only be used to be at the start and ends?
> 
> Something like this:
> 
> Screen Shot 2021-07-01 at 12 17 24 PM

Ah yes, I like this a lot!

File: torch/nn/modules/flatten.py [hunk 3]
```diff
@@ -10,8 +10,9 @@ class Flatten(Module):
     Flattens a contiguous range of dims into a tensor. For use with :class:`~nn.Sequential`.
 
     Shape:
-        - Input: :math:`(N, *dims)`
-        - Output: :math:`(N, \prod *dims)` (for the default case).
+        - Input: :math:`(*, S_{start},..., S_{i}, ..., S_{end}, *)`, where :math:`S_{i}` is the
+          size at dimension :math:`i` and :math:`*` means any number of dimensions including none.
+        - Output: :math:`(*, \prod_{i=start}^{end} S_{i}, *)`
```
Comment (jbschlosser) [inline torch/nn/modules/flatten.py:15]: nit: period at the end here

File: torch/nn/modules/flatten.py [hunk 4]
```diff
@@ -57,8 +58,10 @@ class Unflatten(Module):
       (tuple of `(name, size)` tuples) for `NamedTensor` input.
 
     Shape:
-        - Input: :math:`(N, *dims)`
-        - Output: :math:`(N, C_{\text{out}}, H_{\text{out}}, W_{\text{out}})`
+        - Input: :math:`(*, S_{dim}, *)`. Input can be any shape, where :math:`S_{dim}`
```
Comment (jbschlosser) [inline torch/nn/modules/flatten.py:61]: nit: I think it reads better without the "Input can be any shape" part now but could go either way

File: torch/nn/modules/flatten.py [hunk 5]
```diff
@@ -57,8 +58,10 @@ class Unflatten(Module):
...
+          is the size at dimension :math:`dim`, :math:`S_{dim}=\prod *unflattened\_size`, and
```
Comment (jbschlosser) [inline torch/nn/modules/flatten.py:62]: I realized technically `\prod_i unflattened\_size[i]` is probably more accurate than `\prod *unflattened\_size`.

Alternatively, maybe a `U_{i}` notation for the unflattened size?

**Review by jbschlosser — COMMENTED** [review]
`Flatten` update looks good to me! `Unflatten` is pretty good and I do agree with your thoughts that explicit is better in general.

Only thing I'm wondering is if a `U_{i}` notation will help readability- wdyt? something like:
```
Input: (*, S_{dim}, *), where S_{dim} is the size at dimension dim
Output: (*, U_1, ..., U_n, *), where U = unflattened_size and \prod_i^n U_i = S_{dim}
```

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Review by ssnl — COMMENTED** [review]
Nice! Thanks for fixing. Btw, the formatting will look slightly nicer if you use `\text{start}` and `\text{end}`. No worries if you want to merge this as-is, since this is already drastic improvement.
